src/satellite/merge.ts

```diff
@@ -33,10 +40,37 @@ export const mergeChangesLastWriteWins = (
 }
 
 // Merge the type of two operations so that add wins.
-export const mergeOpTypesAddWins = (first: OpType, second: OpType): OpType => {
+export const mergeOpTypesAddWins = (
+  first: OpType | ChangesOpType,
+  second: OpType | ChangesOpType
+): ChangesOpType => {
   if (first === OPTYPES.delete && second === OPTYPES.delete) {
-    return OPTYPES.delete as OpType
+    return OPTYPES.delete as ChangesOpType
   }
 
-  return OPTYPES.upsert as OpType
+  return OPTYPES.upsert as ChangesOpType
+}
+
+export const mergeOpTags = (
+  local: OplogEntryChanges,
+  remote: OplogEntryChanges
+): Tag[] => {
+  return calculateTag(local.tag, remote.clear_tags, local.clear_tags)
+}
+
+const calculateTag = (tag: Tag | null, tags: Tag[], clear: Tag[]) => {
```

Comment:
We discussed this and I remember agreeing with code, but need to refresh my memory and have things explained to me from your perspective.

When applying incoming operations, we want to add to the local shadow table new incoming tags and make sure that we don't re-add tags that have been cleared by local operations (this is to ensure that a local DELETE after a local INSERT that is being re-deliverred does not re-introduce new tags). Is that what you're doing here?

I would expect that the merge of two entries would generate a new entry with tag being the the union of local and remote tag minus the local clear tags (in this case tag, would have to be a set --- I think this is a valid case: local insert with remote insert). 

ClearTags, I think can always be an union, because they just remove stuff that has been seen.

Somehow, it seems you're doing something different and I need to understand better.

Reply:
I think this is the most critical part of this PR and I want to stop review here before continuing, so let me also make another comment: I think we need more unit tests to test more isolated bits of the code. Current tests are hard to follow for someone coming in. I have presented at least a couple of counter examples, if I am correct in some of my examples, we should have unit tests for showing the issue and use that as inspiration to make more tests.